### PR TITLE
Implement new runtime metrics

### PR DESCRIPTION
**Pull request by dashpole**
Part of https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5655

Changes:

* Move the configuration options to `options.go` without modification.
* Implements the metrics defined here: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/runtime/go-metrics.md.  These are still disabled by default while it is under development.
* Add unit tests

Notes:

It doesn't implement `go.schedule.duration`, as the histogram will need some additional work to figure out.

Based on https://github.com/prometheus/client_golang/pull/955#issue-1090830520, using go's runtime metrics should is more efficient than reading memstats.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add support to configure views when creating MeterProvider using the config package. (#5654)
 - Add log support for the autoexport package. (#5733)
 - Add support for disabling the old runtime metrics using the `OTEL_GO_X_DEPRECATED_RUNTIME_METRICS=false` environment variable. (#5747)
+- Add new runtime metrics to `go.opentelemetry.io/contrib/instrumentation/runtime`, which are still disabled by default. (#5870)
```
Comment (dmathieu) [inline CHANGELOG.md:18]:
```suggestion
- Add new runtime metrics to `go.opentelemetry.io/contrib/instrumentation/runtime`, which are still disabled by default. (#5780)
```

**Review by dmathieu — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

**Comment by MrAlias** [comment]
Waiting to merge so @XSAM can review.

**Review by XSAM — CHANGES_REQUESTED** [review]
Should we add a test for `newCollector` or `refresh` method to ensure the minimum refresh interval can actually work via the collector created by `newCollector`?

File: instrumentation/runtime/runtime.go [hunk 2]
```diff
@@ -97,6 +50,166 @@ func Start(opts ...Option) error {
...
+}
+
+func newCollector(minimumInterval time.Duration) *goCollector {
+	g := &goCollector{
+		sampleBuffer: make([]metrics.Sample, 0, len(runtimeMetrics)),
+		sampleMap:    make(map[string]*metrics.Sample, len(runtimeMetrics)),
+	}
+	for _, runtimeMetric := range runtimeMetrics {
+		g.sampleBuffer = append(g.sampleBuffer, metrics.Sample{Name: runtimeMetric})
+		g.sampleMap[runtimeMetric] = &g.sampleBuffer[len(g.sampleBuffer)-1]
```
[the remark this answers is not in the document]
Comment (dashpole) [reply instrumentation/runtime/runtime.go]: This broke the implementation because we need to reference the position in the slice, not the underlying Sample.

[the remark this answers is not in the document]
Comment (dashpole) [reply instrumentation/runtime/runtime.go]: I've reverted it

**Comment by dashpole** [comment]
Working on tests.  metrics.Read is being strange during unit testing, and seems to always return the same result.

File: instrumentation/runtime/runtime.go [hunk 3]
```diff
@@ -97,6 +50,167 @@ func Start(opts ...Option) error {
...
+
+func newCollector(minimumInterval time.Duration) *goCollector {
+	g := &goCollector{
+		sampleBuffer:    make([]metrics.Sample, 0, len(runtimeMetrics)),
+		sampleMap:       make(map[string]*metrics.Sample, len(runtimeMetrics)),
+		minimumInterval: minimumInterval,
+	}
+	for _, runtimeMetric := range runtimeMetrics {
+		g.sampleBuffer = append(g.sampleBuffer, metrics.Sample{Name: runtimeMetric})
+		g.sampleMap[runtimeMetric] = &g.sampleBuffer[len(g.sampleBuffer)-1]
```
Comment (XSAM) [inline instrumentation/runtime/runtime.go:195]:
```suggestion
		// We are referencing the position of the slice. 
		// Any growth to this `sampleBuffer` slice would break the implementation.
		g.sampleBuffer = append(g.sampleBuffer, metrics.Sample{Name: runtimeMetric})
		g.sampleMap[runtimeMetric] = &g.sampleBuffer[len(g.sampleBuffer)-1]
```

Should we add a comment here to declare that we are referencing the position of the slice and that any growth to this `sampleBuffer` slice would break the implementation?

Comment (dashpole) [reply instrumentation/runtime/runtime.go]: added a similar disclaimer.

**Comment by dashpole** [comment]
I had to add code that allocates memory to ensure the metric actually changes for the refresh test.  I ran it locally with `-count 10000`, and it passes, so hopefully it won't be flaky.

**Comment by dashpole** [comment]
`/gc/heap/allocs:bytes` and `/gc/heap/allocs:objects` don't seem to be populated on windows.  I'm investigating

**Comment by dashpole** [comment]
I think I got the test passing

File: instrumentation/runtime/test/runtime_test.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+			t.Fatalf("unexpected data type %v", a)
+		}
+	}
+}
+
+func allocateMemory(buffer [][]byte) {
+	newBuffer := make([]byte, 1000000)
+	for i := range newBuffer {
+		newBuffer[i] = 0
+	}
```
Comment (MrAlias) [inline instrumentation/runtime/test/runtime_test.go:159]: Go does this by default for all slices created with `make`.

```suggestion
```

File: instrumentation/runtime/test/runtime_test.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	buffer = append(buffer, newBuffer)
+}
```
Comment (MrAlias) [inline instrumentation/runtime/test/runtime_test.go:161]: The passed slice is never used again. The GC will collect the allocated space right after this returns, right?

Maybe pass a pointer, return the slice, or just replace this call with `buffer append(buffer, make([]byte, 1000000))`

File: instrumentation/runtime/test/runtime_test.go (hunk 4 above)
Comment (dashpole) [reply instrumentation/runtime/test/runtime_test.go]: removed.

File: instrumentation/runtime/test/runtime_test.go (hunk 5 above)
Comment (dashpole) [reply instrumentation/runtime/test/runtime_test.go]: returned the slice.

**Review by MrAlias — APPROVED** [review]

**Comment by MrAlias** [comment]
Looks like `go mod tidy` needs to be run.

**Comment by MrAlias** [comment]
@XSAM PTAL.

**Comment by MrAlias** [comment]
@XSAM planning to merge this in an hour or two unless I hear from you otherwise.

**Review by XSAM — APPROVED** [review]
